### PR TITLE
Nicer C++ stuff for motion metamodule/utils

### PR DESCRIPTION
**Pull request by KipHamiltons**
`const, constexpr, [[nodiscard]]` are cool. We should use them more.

### CODE REVIEW

File: shared/utility/motion/splines/TrajectoryUtils.hpp [hunk 1]
```diff
@@ -87,52 +88,52 @@ namespace utility::motion::splines {
...
+        [[nodiscard]] constexpr operator Value() const {
             return value;
         }
-        operator uint8_t() const {
+        [[nodiscard]] constexpr operator uint8_t() const {
             return value;
         }
 
-        operator std::string() const {
+        [[nodiscard]] operator std::string() const {
```
Comment (CMurtagh-LGTM) [inline shared/utility/motion/splines/TrajectoryUtils.hpp:136]: Definition in `.cpp` please

**Review by CMurtagh-LGTM — CHANGES_REQUESTED** [review]

File: module/motion/QuinticWalk/src/WalkEngine.hpp [hunk 2]
```diff
@@ -107,16 +107,15 @@ namespace module::motion {
      */
     class QuinticWalkEngine {
     public:
-        /**
-         * Initialization
-         */
-        QuinticWalkEngine();
+        QuinticWalkEngine() {
+            trajectoriesInit(trajs);
+        };
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/WalkEngine.hpp:112]: Why? It's one line. I think it's a more readable experience having it in the header

File: module/motion/QuinticWalk/src/WalkEngine.hpp [hunk 3]
```diff
@@ -125,28 +124,28 @@ namespace module::motion {
...
-        float getTrajsTime() const {
+        [[nodiscard]] constexpr float getTrajsTime() const {
             return phase < 0.5f ? phase / params.freq : (phase - 0.5f) / params.freq;
         }
 
         /**
          * Get the footstep object.
          */
-        Footstep getFootstep() {
+        [[nodiscard]] Footstep getFootstep() {
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply module/motion/QuinticWalk/src/WalkEngine.hpp:134]: functions defined inside classes are implicitly inline

File: shared/utility/motion/splines/Combination.hpp [hunk 4]
```diff
@@ -66,19 +66,19 @@ namespace utility::motion::splines {
...
             }
         }
 
         /**
          * Return the next combination.
-         * Return empty sdt::vector when iteration is finished
+         * Return empty std::vector when iteration is finished
          */
-        std::vector<size_t> nextCombination() {
+        [[nodiscard]] std::vector<size_t> nextCombination() {
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply shared/utility/motion/splines/Combination.hpp:81]: functions defined inside classes are implicitly inline

File: shared/utility/motion/splines/Combination.hpp [hunk 5]
```diff
@@ -101,26 +101,26 @@ namespace utility::motion::splines {
...
+        std::vector<size_t> indexes{};
+        size_t n = 0;
+        size_t k = 0;
 
         /**
          * Increment by one the indexes container at digit i (recursively).
          * Return true on iteration end.
          */
-        bool incrIndexes(size_t i) {
+        bool incrIndexes(const size_t& i) {
```
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply shared/utility/motion/splines/Combination.hpp:117]: functions defined inside classes are implicitly inline

File: shared/utility/motion/splines/TrajectoryUtils.hpp (hunk 1 above)
Comment (KipHamiltons) [reply shared/utility/motion/splines/TrajectoryUtils.hpp:136]: why?

Comment (TrentHouliston) [reply shared/utility/motion/splines/TrajectoryUtils.hpp:136]: ODR

File: shared/utility/motion/splines/Combination.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (TrentHouliston) [reply shared/utility/motion/splines/Combination.hpp:117]: They are most certainly not implicitly inline, the constructor is, but everything else isn't. Unless it's a template

[the remark this answers is not in the document]
Comment (TrentHouliston) [reply shared/utility/motion/splines/Combination.hpp:117]: Although that said I'm having trouble finding documents that state either way

[the remark this answers is not in the document]
Comment (KipHamiltons) [reply shared/utility/motion/splines/Combination.hpp:117]: Yikes. well then. play on

File: shared/utility/motion/splines/TrajectoryUtils.hpp (hunk 1 above)
Comment (KipHamiltons) [reply shared/utility/motion/splines/TrajectoryUtils.hpp:136]: Granted, you have showed me why that's true for the general case, but this case is for a "smart enum" struct which doesn't even have its own hpp. I suppose I'll make it inline?

File: shared/utility/motion/splines/Combination.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (KipHamiltons) [reply shared/utility/motion/splines/Combination.hpp:117]: I was definitely thinking of templates, FWIW

**Review by CMurtagh-LGTM — DISMISSED** [review]
LGTM

**Review by CMurtagh-LGTM — APPROVED** [review]
LGTM

**Comment by ysims** [comment]
Tested and looks like it's fine.

**Pull request merged**